### DISCUSSION TITLE
Training Fastconformer-CTC on another language.

### DISCUSSION THREAD

**Discussion by bharathraj-v** [NVIDIA-NeMo/NeMo#8123] [Q&A] [answered]
Hi,
I'm new to NeMo and looking to train the Fastconformer-CTC model on Telugu datasets. Telugu is an Indian language which is highly inflectional and somewhat agglutinative. I have around 800 hours of data, what's the best approach to achieve good results here? Do I fine-tune a pre-trained model or train one from scratch?

Also, what's the config that's best suited for this case? And, do I go for subword tokenization for a language such as this or does the standard bpe tokenizer provided in the scripts suffice? Any guidance regarding this would be of great help

Thanks!

**Comment by titu1994** [comment]
@nithinraok wrote this post for a Telegu speech model here - https://blogs.nvidia.com/blog/speech-ai-telugu-language-breakthrough/

He could provide you with specific advice

**Reply by titu1994** [reply, marked as answer]
Overall yes, use subwords (spe unigram or just spe bpe). Config - Fast Conformer large - try to finetune and not train from scratch, it will always lead to better results. The ASR CTC finetune  tutorial has most of the tricks for fine-tuning that @nithinraok applied

**Reply by bharathraj-v** [reply]
Thanks!

**Reply by nithinraok** [reply]
Yes as @titu1994 mentioned, create a spe bpe based tokenizer from telugu words and finetune a XL/L En ASR model. 
L: https://huggingface.co/nvidia/stt_en_fastconformer_ctc_large 
XL: https://huggingface.co/nvidia/stt_en_fastconformer_ctc_xlarge

**Reply by bharathraj-v** [reply]
Okay, thanks! So for Telugu's spe bpe tokenizer, what's the ideal vocab size I should go for? @titu1994 @nithinraok

**Reply by nithinraok** [reply]
You may experiment with various sizes (256,512, 1024). 256 should be good enough based on our prior experiments.

**Reply by bharathraj-v** [reply]
I've been trying to fine-tune stt_en_fastconformer_ctc_large with Telugu data and even after 6 epochs, the model shows no sign of improvement. I wasn't expecting it to converge but during the validation, the model is just outputting one character constantly and did not improve. I don't know what I'm doing wrong here. Any guidance regarding this?
![image](https://github.com/NVIDIA/NeMo/assets/69118968/81208b97-749b-4d80-87cb-dfcdec1487af)
```
import nemo.collections.asr as nemo_asr
import copy
from omegaconf import OmegaConf, open_dict
import torch
import pytorch_lightning as ptl

model = nemo_asr.models.ASRModel.from_pretrained("nvidia/stt_en_fastconformer_ctc_large")
model.change_vocabulary(new_tokenizer_dir='resources/tokenizer_spe_bpe_v1024', new_tokenizer_type='bpe')

cfg = copy.deepcopy(model.cfg)

with open_dict(cfg):
    cfg.train_ds.manifest_filepath = 'resources/final_te_manifests/train_manifest.json'
    cfg.test_ds.manifest_filepath = 'resources/final_te_manifests/test_manifest.json'
    cfg.validation_ds.manifest_filepath = 'resources/final_te_manifests/dev_manifest.json'
    cfg.optim.lr = 0.9
    cfg.batch_size=19
    cfg.train_ds.batch_size=19
    cfg.test_ds.batch_size=19
    cfg.validation_ds.batch_size=19

model.setup_training_data(cfg.train_ds)
model.setup_multiple_validation_data(cfg.validation_ds)
model.setup_multiple_test_data(cfg.test_ds)

model.cfg = cfg

EPOCHS = 25 

torch.set_float32_matmul_precision('medium')
torch.cuda.is_available()

trainer = ptl.Trainer(devices=4,
                      accelerator='gpu',
                      max_epochs=EPOCHS,
                      enable_checkpointing=True,
                      logger=True,
                      log_every_n_steps=50,
                      check_val_every_n_epoch=1,
                      strategy = 'ddp' 
                      )


# Setup model with the trainer
model.set_trainer(trainer)

# Finally, update the model's internal config
model.cfg = model._cfg

trainer.fit(model)

model.save_to("resources/fastconformer-telugu-2.nemo")
```



This is the script I used where I kept most of the default values and did very little modifications. I'm very new to training ASR models, so any guidance regarding this would really be of massive help.

**Reply by titu1994** [reply]
The learning rate is increasing all the way to 1.0 that is simply too high to train any model. Learning rate when Fine-tuning should not go above 5e-4

**Reply by bharathraj-v** [reply]
@titu1994, Thanks! One more thing, what's the ideal min_lr in this case?
The optimizer config is currently:
```
{
 'name': 'adamw',
 'lr': 0.0005, 
 'betas': [0.9, 0.98], 
 'weight_decay': 0.001, 
 'sched': {
    'name': 'CosineAnnealing', 
    'warmup_steps': 15000, 
    'warmup_ratio': None, 
    'min_lr': 0.0001
    }
}
```
Will this config work fine or are any changes needed?

**Reply by nithinraok** [reply]
1e-6 should be good enough

**Reply by bharathraj-v** [reply]
Okay, thank you so much!

**Reply by nithinraok** [reply]
Try experimenting with warmup_steps (5000,10000, 15000) and max_steps(aka epochs if you are relying on epochs) as well.

**Reply by bharathraj-v** [reply]
Okay! Also to verify, my understanding of how the LR scheduling process will go is - 
1. stage 1: warmup linearly increases the LR till it hits max
2. stage 2: cosine annealing decreases and increases the lr acc. to a cosine curve

Weight decay is separate from this and decreases the model's weights to prevent overfitting. Am I correct in this understanding? I just want to make sure
